### PR TITLE
From Ogre 2.1 to Ogre 2.2

### PR DESCRIPTION
**Pull request by ahcorde**
This PR updates the Ogre2 from Ogre 2.1 to Ogre 2.2.

There are some issues that need to be resolved yet 

 - GPURays not working
 - When more than one rendering sensor is added the cameras are not showing anything
 - The 3DScene is completely black ( not getting the right GLID )
 - Expose EGL
 - Clean up code

Related with:
 - https://github.com/ignition-forks/ogre-next/pull/1
 - https://github.com/ignitionrobotics/ign-cmake/pull/157

### CODE REVIEW

**Comment by ahcorde** [comment]
@mjcarroll and @iche033, this PR is ready for a initial review. 

There is a know issue in the Ogre2SelectionBuffer but the rest of classes should work fine:

 - Camera
 - DepthCamera
 - ThermalCamera
 - GPURay
 - LidarVisual
 - Lightvisual
 - RayQuery 
 - Grid
 - Renderpass
 - Gizmo

**Comment by ahcorde** [comment]
I fixed the Ogre2Selectionbuffer.

I will create a PR on top of this one to enable EGL

**Comment by mjcarroll** [comment]
Can you split the `media` updates into their own commit to make it a bit easier to review?

File: ogre2/src/Ogre2ThermalCamera.cc [hunk 1]
```diff
@@ -656,7 +647,7 @@ void Ogre2ThermalCamera::CreateThermalTexture()
   // compositor_node ThermalCamera
   // {
   //   in 0 rt_input
-  //   texture depthTexture target_width target_height PF_D32_FLOAT
+  //   texture depthTexture target_width target_height PFG_D32_FLOAT
   //   texture colorTexture target_width target_height PF_R8G8B8A8
```
Comment (mjcarroll) [inline ogre2/src/Ogre2ThermalCamera.cc:651]: This type needs to be updated as well

File: ogre2/src/Ogre2RenderTarget.cc [hunk 2]
```diff
@@ -309,6 +300,12 @@ void Ogre2RenderTarget::RebuildCompositor()
 //////////////////////////////////////////////////
 void Ogre2RenderTarget::Copy(Image &_image) const
 {
+  Ogre::TextureGpu * texture = this->RenderTarget();
+
+  if (texture == nullptr){
+    return;
+  }
+
   // TODO(anyone) handle Bayer conversions
```
Comment (mjcarroll) [inline ogre2/src/Ogre2RenderTarget.cc:309]: This may be handed by `Ogre::Image2` if I understand correctly?

**Review by mjcarroll — COMMENTED** [review]
A few comments, moving on to checking tests and examples

File: ogre2/src/Ogre2RenderEngine.cc [hunk 3]
```diff
@@ -340,6 +350,8 @@ void Ogre2RenderEngine::CreateContext()
 
   if (!this->dummyDisplay)
   {
+    // Not able to create a Xwindow, try to run in headless mode
+    SetHeadless(true);
```
Comment (iche033) [inline ogre2/src/Ogre2RenderEngine.cc:354]: `this->SetHeadless`

File: ogre2/src/Ogre2DepthCamera.cc [hunk 4]
```diff
@@ -589,119 +589,124 @@ void Ogre2DepthCamera::CreateDepthTexture()
...
     rt0TexDef->numMipmaps = 0;
     rt0TexDef->widthFactor = 1;
     rt0TexDef->heightFactor = 1;
-    rt0TexDef->formatList = {Ogre::PF_FLOAT32_RGBA};
-    rt0TexDef->fsaa = 0;
-    rt0TexDef->uav = false;
-    rt0TexDef->automipmaps = false;
-    rt0TexDef->hwGammaWrite = Ogre::TextureDefinitionBase::BoolFalse;
+    rt0TexDef->format = Ogre::PFG_RGBA32_FLOAT;
+    rt0TexDef->textureFlags |= !Ogre::TextureFlags::Uav;
```
Comment (iche033) [inline ogre2/src/Ogre2DepthCamera.cc:600]: there is a compile warning here and other places with this line. I think it's for removing the flag? i.e. `rt0TexDef->textureFlags &= ~Ogre::TextureFlags::Uav;`? or maybe we can just set it to `Ogre::TextureFlags::RenderToTexture`

**Review by iche033 — COMMENTED** [review]
did a first pass review. Looking good

**Comment by chapulina** [comment]
After some internal discussion, we decided to target this at Fortress. Removing from Edifice's beta.

**Comment by ahcorde** [comment]
there are still two issues:

 - sky is a little bit darker in ign-gazebo, one thing to take in account it's that the camera looks right. Probably it's something related with the gamma correction in ign-gazebo, we are applying the gamma correction [here](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/src/gui/plugins/scene3d/Scene3D.cc#L2490-L2491) but for some reason it's not affecting the background.

Selección_120

 - lightmaps are red

2021-03-25T17_03_05 548978603

**Comment by iche033** [comment]
the remaining two issues with red colors from light map (and albedo map), and dark sky texture should be fixed in 62bb629

**Comment by iche033** [comment]
fixed toggling sky 2e9d62e

tests are all passing now.

**Comment by ahcorde** [comment]
I made some changes in Ogre2Material to fix the memory leak issue. @iche033 or @mjcarroll do you mind to test it locally ?

You should compile `ign-rendering` in `Debug` mode and run `heaptrack`.

```bash
heaptrack -o output ruby <path to ignition> gazebo -v 4 -s -r  camera_sensor.sdf
```

Then add and remove some models. Finally 

```bash
heaptrack -a output.gz
```

**Comment by iche033** [comment]
here's what I get from running heaptrack on ign-gazebo gui and repeatedly spawning and removing models:

ign-rendering-ogre2 2_heaptrack

Memory looks stable

**Comment by darksylinc** [comment]
I added a PR #354 against this PR because I tested it (well done! it worked out of the box) and found some unusual warnings in stdout.

These were cause by the unwanted files that **_must_** be removed. Please note that after updating, you must remove them from your installation folder as well (`make install` won't remove them). They're usually located in ` ign/install/share/ignition/ignition-rendering6/ogre2/media/Hlms`. Just remove the folder and type `make install` again.

**Comment by mjcarroll** [comment]
@osrf-jenkins retest this please

Checking Windows.

**Comment by chapulina** [comment]
Windows failed:

```
C:\vcpkg\installed\x64-windows\include\OGRE-2.2\Hlms/Pbs/OgreHlmsPbsDatablock.h(38,14): fatal error C1083: Cannot open include file: 'OgreHlmsTextureBaseClass.h': No such file or directory (compiling source file C:\Jenkins\workspace\ign_rendering-pr-win\ws\ign-rendering\ogre2\src\Ogre2DynamicRenderable.cc) [C:\Jenkins\workspace\ign_rendering-pr-win\ws\build\ignition-rendering6\ogre2\src\ignition-rendering6-ogre2.vcxproj]
```

**Comment by mjcarroll** [comment]
Looks like it may be missing `installed/x64-windows/include/OGRE-2.2/Hlms/Common` as an include path.  The file is definitely there.  I'll see what I can do.

**Comment by mjcarroll** [comment]
@osrf-jenkins retest this please

**Comment by j-rivero** [comment]
Not sure if this expected at this point or excuse me if it is out of the scope from this PR, looking [into the logs](https://build.osrfoundation.org/job/ignition_rendering-ci-pr_any-homebrew-amd64/1708/consoleText), found some problems on Brew:

```cmake
-- -- Finding OGRE 2.2
-- Looking for IgnOGRE2 - not found

CUDA_TOOLKIT_ROOT_DIR not found or specified
-- Could NOT find CUDA (missing: CUDA_TOOLKIT_ROOT_DIR CUDA_NVCC_EXECUTABLE CUDA_INCLUDE_DIRS CUDA_CUDART_LIBRARY) 
-- optix library not found.  Please locate before proceeding.
-- OptiX headers (optix.h and friends) not found.  Please locate before proceeding.
-- optix Prime library not found.  Please locate before proceeding.
-- Could NOT find OptiX (missing: OptiX_FOUND) (Required is at least version "3.8.0")
-- Looking for OptiX - not found

CMake Warning at /usr/local/share/cmake/ignition-cmake2/cmake2/IgnConfigureBuild.cmake:53 (message):
   CONFIGURATION WARNINGS:
   -- Skipping component [ogre2]: Missing dependency [IgnOGRE2] (Components: HlmsPbs, HlmsUnlit, Overlay).
      ^~~~~ Set SKIP_ogre2=true in cmake to suppress this warning.
```

**Comment by darksylinc** [comment]
INTEGRATION_lidar_visual is passing however there's lot of:

> [Err] [Ogre2GpuRays.cc:262] Error casting user data: Unexpected index

Which aren't happening in Edifice. I'm taking a look to see what's going on.

**Review by mjcarroll — APPROVED** [review]

**Comment by chapulina** [comment]
I ticketed https://github.com/ignition-tooling/release-tools/issues/509 so the macOS and Windows builds to turn yellow when Ogre 2 isn't found.

For macOS, we may need to set `SKIP_ogre2` for now. But we shouldn't regress on Windows CI. What's the status of that?

**Comment by mjcarroll** [comment]
The OGRE2 rendering plugin builds correctly on Windows with OGRE2.2, but tests fail with: 

`Failed to load plugin ignition-rendering-ogre2: couldn't load library on path`, which is the same as with OGRE2.1, so not a regression, but not an improvement either.

**Comment by chapulina** [comment]
Ahh I hadn't noticed that Windows found Ogre2, all good!

@iche033 , can we get this in?

**Review by iche033 — APPROVED** [review]

**Comment by iche033** [comment]
yes merging! :crossed_fingers:

**Pull request merged**